src/Features/Core/Portable/NameTupleElement/AbstractNameTupleElementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.NameTupleElement
+{
+    abstract class AbstractNameTupleElementCodeRefactoringProvider<TArgumentSyntax, TTupleExpressionSyntax> : CodeRefactoringProvider
+        where TArgumentSyntax : SyntaxNode
+        where TTupleExpressionSyntax : SyntaxNode
+    {
+        protected abstract bool IsCloseParenOrComma(SyntaxToken token);
+        protected abstract TArgumentSyntax WithName(TArgumentSyntax argument, string argumentName);
+        protected abstract bool IsUnnamedElement(TArgumentSyntax node);
+        protected abstract SeparatedSyntaxList<TArgumentSyntax> GetArguments(TTupleExpressionSyntax tuple);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            if (context.Span.Length > 0)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var position = context.Span.Start;
+            var token = root.FindToken(position);
+            if (token.Span.Start == position &&
+                IsCloseParenOrComma(token))
+            {
+                token = token.GetPreviousToken();
+                if (token.Span.End != position)
+                {
+                    return;
+                }
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var argument = root.FindNode(token.Span)
+                .GetAncestorsOrThis<TArgumentSyntax>()
+                .FirstOrDefault(node => syntaxFacts.IsTupleExpression(node.Parent));
```

Comment:
This seems odd. Seems like this would work for an argument of an invocation deep inside a tuple. Shouldn't we just get the first TArgumentSyntax and bail out if its parent is not a tuple?

Reply:
if you have `(element1, Method(arg1, arg2))` does the refactoring trigger on the comma after `arg1`?

Reply:
Added test.